flax/struct.py

```diff
@@ -162,11 +164,50 @@ def from_state_dict(x, state):
   serialization.register_serialization_state(
       data_clz, to_state_dict, from_state_dict)
 
+  def new_setattr(obj, name, value):
+    if obj.__mutable__:
+      object.__setattr__(obj, name, value)
+    else:
+      raise dataclasses.FrozenInstanceError(f"cannot assign to field '{name}'")
+
+  data_clz.__mutable__ = False # denotes whether the object's currently mutable
+  data_clz.__setattr__ = new_setattr # type: ignore
+
   # add a _flax_dataclass flag to distinguish from regular dataclasses
   data_clz._flax_dataclass = True # type: ignore[attr-defined]
 
   return data_clz # type: ignore
 
+@contextlib.contextmanager
+def create_mutable_copy(obj):
```

Comment:
Hi @chiamp!

Would you be open to adding a type annotation here, something like:
```
from typing import Generator

@contextlib.contextmanager
def create_mutable_copy(obj: _T) -> Generator[_T, None, None]:
  ...
```

It'd be nice for tab completion on the yielded pytree :slightly_smiling_face: 

I added an inner function in `jax_dataclasses` for a similar purpose, but it looks like that's not actually necessary (probably the `contextlib` stubs have improved a lot since then?).

Reply:
Thanks for the suggestion @brentyi! I added it in